syntax = "proto3";

package aifs.v1;

// Asset kinds
enum AssetKind {
  BLOB = 0;
  TENSOR = 1;
  EMBED = 2;
  ARTIFACT = 3;
}

// Asset metadata
message AssetMetadata {
  string asset_id = 1;
  AssetKind kind = 2;
  int64 size = 3;
  string created_at = 4;
  map<string, string> metadata = 5;
}

// Parent edge for lineage tracking
message ParentEdge {
  string parent_asset_id = 1;
  string transform_name = 2;
  string transform_digest = 3;
}

// Asset chunk for streaming
message Chunk {
  bytes data = 1;
}

// Put asset request
message PutAssetRequest {
  AssetKind kind = 1;
  map<string, string> metadata = 2;
  repeated ParentEdge parents = 3;
  bytes embedding = 4;  // Optional embedding vector
  repeated Chunk chunks = 5;  // Streamed chunks
}

// Put asset response
message PutAssetResponse {
  string asset_id = 1;
}

// Get asset request
message GetAssetRequest {
  string asset_id = 1;
  bool include_data = 2;  // Whether to include the actual data
}

// Get asset response
message GetAssetResponse {
  AssetMetadata metadata = 1;
  repeated ParentEdge parents = 2;
  repeated string children = 3;  // Child asset IDs
  bytes data = 4;  // Only included if include_data is true
  string uri = 5;  // aifs:// URI for the asset
}

// Vector search request
message VectorSearchRequest {
  bytes query_embedding = 1;
  int32 k = 2;  // Number of results to return
  map<string, string> filter = 3;  // Optional metadata filters
}

// Vector search result
message SearchResult {
  string asset_id = 1;
  float score = 2;
  AssetMetadata metadata = 3;
}

// Vector search response
message VectorSearchResponse {
  repeated SearchResult results = 1;
}

// List assets request
message ListAssetsRequest {
  int32 limit = 1;  // Maximum number of assets to return
  int32 offset = 2;  // Number of assets to skip
}

// List assets response
message ListAssetsResponse {
  repeated AssetMetadata assets = 1;
}

// Subscribe events request
message SubscribeEventsRequest {
  string filter = 1;  // Optional filter for events
  bool include_lineage = 2;  // Include lineage events
  bool include_drift = 3;  // Include drift events
}

// Event message
message Event {
  string event_id = 1;
  string event_type = 2;  // "asset_created", "asset_updated", "lineage_added", etc.
  string asset_id = 3;
  string namespace = 4;
  int64 timestamp = 5;
  map<string, string> metadata = 6;
}

// Subscribe events response
message SubscribeEventsResponse {
  repeated Event events = 1;
}

// Error response
message ErrorResponse {
  int32 code = 1;  // gRPC status code
  string reason = 2;
  string detail = 3;
}

// Create snapshot request
message CreateSnapshotRequest {
  string namespace = 1;
  repeated string asset_ids = 2;
  map<string, string> metadata = 3;
}

// Create snapshot response
message CreateSnapshotResponse {
  string snapshot_id = 1;
  string merkle_root = 2;
}

// Get snapshot request
message GetSnapshotRequest {
  string snapshot_id = 1;
}

// Get snapshot response
message GetSnapshotResponse {
  string snapshot_id = 1;
  string namespace = 2;
  string merkle_root = 3;
  string created_at = 4;
  map<string, string> metadata = 5;
  repeated string asset_ids = 6;
}

// Delete asset request
message DeleteAssetRequest {
  string asset_id = 1;
  bool force = 2;  // Force deletion even if referenced by other assets
}

// Delete asset response
message DeleteAssetResponse {
  bool success = 1;
  string message = 2;
}

// List namespaces request
message ListNamespacesRequest {
  int32 limit = 1;
  int32 offset = 2;
}

// List namespaces response
message ListNamespacesResponse {
  repeated NamespaceInfo namespaces = 1;
}

// Namespace information
message NamespaceInfo {
  string namespace_id = 1;
  string name = 2;
  string description = 3;
  string created_at = 4;
  map<string, string> metadata = 5;
}

// Get namespace request
message GetNamespaceRequest {
  string namespace_id = 1;
}

// Get namespace response
message GetNamespaceResponse {
  NamespaceInfo namespace = 1;
}

// Verify asset request
message VerifyAssetRequest {
  string asset_id = 1;
}

// Verify asset response
message VerifyAssetResponse {
  bool valid = 1;
  string message = 2;
  string computed_hash = 3;
  string stored_hash = 4;
}

// Verify snapshot request
message VerifySnapshotRequest {
  string snapshot_id = 1;
  bytes public_key = 2;  // Optional public key for signature verification
}

// Verify snapshot response
message VerifySnapshotResponse {
  bool valid = 1;
  string message = 2;
  string merkle_root = 3;
  bool signature_valid = 4;
}

// Create branch request
message CreateBranchRequest {
  string branch_name = 1;
  string namespace = 2;
  string snapshot_id = 3;
  map<string, string> metadata = 4;
}

// Create branch response
message CreateBranchResponse {
  bool success = 1;
  string message = 2;
}

// Get branch request
message GetBranchRequest {
  string branch_name = 1;
  string namespace = 2;
}

// Get branch response
message GetBranchResponse {
  string branch_name = 1;
  string namespace = 2;
  string snapshot_id = 3;
  string created_at = 4;
  string updated_at = 5;
  map<string, string> metadata = 6;
}

// List branches request
message ListBranchesRequest {
  string namespace = 1;  // Optional namespace filter
  int32 limit = 2;
}

// List branches response
message ListBranchesResponse {
  repeated GetBranchResponse branches = 1;
}

// Delete branch request
message DeleteBranchRequest {
  string branch_name = 1;
  string namespace = 2;
}

// Delete branch response
message DeleteBranchResponse {
  bool success = 1;
  string message = 2;
}

// Get branch history request
message GetBranchHistoryRequest {
  string branch_name = 1;
  string namespace = 2;
  int32 limit = 3;
}

// Branch history entry
message BranchHistoryEntry {
  int32 id = 1;
  string branch_name = 2;
  string namespace = 3;
  string old_snapshot_id = 4;
  string new_snapshot_id = 5;
  string updated_at = 6;
  map<string, string> metadata = 7;
}

// Get branch history response
message GetBranchHistoryResponse {
  repeated BranchHistoryEntry history = 1;
}

// Create tag request
message CreateTagRequest {
  string tag_name = 1;
  string namespace = 2;
  string snapshot_id = 3;
  map<string, string> metadata = 4;
}

// Create tag response
message CreateTagResponse {
  bool success = 1;
  string message = 2;
}

// Get tag request
message GetTagRequest {
  string tag_name = 1;
  string namespace = 2;
}

// Get tag response
message GetTagResponse {
  string tag_name = 1;
  string namespace = 2;
  string snapshot_id = 3;
  string created_at = 4;
  map<string, string> metadata = 5;
}

// List tags request
message ListTagsRequest {
  string namespace = 1;  // Optional namespace filter
  int32 limit = 2;
}

// List tags response
message ListTagsResponse {
  repeated GetTagResponse tags = 1;
}

// Delete tag request
message DeleteTagRequest {
  string tag_name = 1;
  string namespace = 2;
}

// Delete tag response
message DeleteTagResponse {
  bool success = 1;
  string message = 2;
}

// AIFS service
service AIFS {
  // Store an asset
  rpc PutAsset(stream PutAssetRequest) returns (PutAssetResponse);
  
  // Retrieve an asset
  rpc GetAsset(GetAssetRequest) returns (GetAssetResponse);
  
  // Delete an asset
  rpc DeleteAsset(DeleteAssetRequest) returns (DeleteAssetResponse);
  
  // List assets
  rpc ListAssets(ListAssetsRequest) returns (ListAssetsResponse);
  
  // Search for similar assets
  rpc VectorSearch(VectorSearchRequest) returns (VectorSearchResponse);
  
  // Create a snapshot
  rpc CreateSnapshot(CreateSnapshotRequest) returns (CreateSnapshotResponse);
  
  // Get a snapshot
  rpc GetSnapshot(GetSnapshotRequest) returns (GetSnapshotResponse);
  
  // Subscribe to events
  rpc SubscribeEvents(SubscribeEventsRequest) returns (stream SubscribeEventsResponse);
  
  // List namespaces
  rpc ListNamespaces(ListNamespacesRequest) returns (ListNamespacesResponse);
  
  // Get namespace
  rpc GetNamespace(GetNamespaceRequest) returns (GetNamespaceResponse);
  
  // Verify asset integrity
  rpc VerifyAsset(VerifyAssetRequest) returns (VerifyAssetResponse);
  
  // Verify snapshot integrity and signature
  rpc VerifySnapshot(VerifySnapshotRequest) returns (VerifySnapshotResponse);
  
  // Branch management
  rpc CreateBranch(CreateBranchRequest) returns (CreateBranchResponse);
  rpc GetBranch(GetBranchRequest) returns (GetBranchResponse);
  rpc ListBranches(ListBranchesRequest) returns (ListBranchesResponse);
  rpc DeleteBranch(DeleteBranchRequest) returns (DeleteBranchResponse);
  rpc GetBranchHistory(GetBranchHistoryRequest) returns (GetBranchHistoryResponse);
  
  // Tag management
  rpc CreateTag(CreateTagRequest) returns (CreateTagResponse);
  rpc GetTag(GetTagRequest) returns (GetTagResponse);
  rpc ListTags(ListTagsRequest) returns (ListTagsResponse);
  rpc DeleteTag(DeleteTagRequest) returns (DeleteTagResponse);
}

// Health service
service Health {
  rpc Check(HealthCheckRequest) returns (HealthCheckResponse);
}

message HealthCheckRequest {}
message HealthCheckResponse {
  bool healthy = 1;
  string status = 2;
}

// Introspect service
service Introspect {
  rpc GetInfo(IntrospectRequest) returns (IntrospectResponse);
}

message IntrospectRequest {}
message IntrospectResponse {
  string version = 1;
  string config = 2;
  repeated string features = 3;
}

// Admin service
service Admin {
  rpc CreateNamespace(CreateNamespaceRequest) returns (CreateNamespaceResponse);
  rpc PruneSnapshot(PruneSnapshotRequest) returns (PruneSnapshotResponse);
  rpc ManagePolicy(ManagePolicyRequest) returns (ManagePolicyResponse);
}

message CreateNamespaceRequest {
  string name = 1;
}
message CreateNamespaceResponse {
  bool success = 1;
  string namespace_id = 2;
}
message PruneSnapshotRequest {
  string snapshot_id = 1;
}
message PruneSnapshotResponse {
  bool success = 1;
}
message ManagePolicyRequest {
  string namespace_id = 1;
  string policy = 2;
}
message ManagePolicyResponse {
  bool success = 1;
}

// Metrics service
service Metrics {
  rpc GetMetrics(MetricsRequest) returns (MetricsResponse);
}

message MetricsRequest {}
message MetricsResponse {
  string prometheus_metrics = 1;
  string opentelemetry_metrics = 2;
}

// Format service
service Format {
  rpc FormatStorage(FormatRequest) returns (FormatResponse);
}

message FormatRequest {
  bool dry_run = 1;
}
message FormatResponse {
  bool success = 1;
  string root_snapshot_id = 2;
  string log = 3;
}